flocker/node/agents/test/test_blockdevice.py

```diff
@@ -5249,3 +5257,231 @@ def test_get_device_path_until_detach(self):
+                cluster_state_callback(cluster_state)
+                r = deployer.calculate_changes(
+                    desired_configuration, cluster_state, local_state)
+                return r
+            d.addCallback(got_changes)
+
+            def run_change_if_needed(change):
+                """
+                Returns True if the change was needed.
+                """
+                if change == NoOp():
+                    return False
+                else:
+                    state_changing = run_state_change(change, deployer)
+                    state_changing.addCallback(lambda _: True)
+                    return state_changing
+            d.addCallback(run_change_if_needed)
+            return d
+
+        def loop_until_converged():
+            result = converge()
+
+            def repeat_if_necessary(changes_needed):
+                if changes_needed:
+                    return loop_until_converged()
+                else:
+                    return None
+
+            result.addCallback(repeat_if_necessary)
+            return result
+        return loop_until_converged()
+
+    def test_protect_against_docker(self):
+        """
+        Verify that in a standard lifecycle of a dataset there is never a time
+        after the path of a dataset is reported to the control node that a
+        successful write to that path is not persisted to the backing loop
+        device.
+        """
+        evaluate_function = None
+
+        def callback(when_description):
+            if evaluate_function:
+                evaluate_function(when_description)
+        actual_blockdevice_manager = BlockDeviceManager()
+        proxy_blockdevice_manager = CallbackProxy(
+            actual_blockdevice_manager, callback)
+
+        # Set up a deployer with a blockdevice_manager that has hooks so we can
+        # inject code around interactions with the blockdevices on the system.
+        deployer = create_blockdevicedeployer(
+            self, node_uuid=self.NODE_UUID,
+            block_device_manager=proxy_blockdevice_manager)
+
+        testing_mountroot = mountroot_for_test(self)
+
+        def has_file_named(blockdevice, filename):
+            """
+            Returns whether a given blockdevice has a file with the given name.
+            """
+            test_mountpoint = testing_mountroot.child(str(uuid4()))
+            test_mountpoint.makedirs()
+            actual_blockdevice_manager.mount(blockdevice, test_mountpoint)
+            result = test_mountpoint.child(filename).exists()
+            actual_blockdevice_manager.unmount(blockdevice)
+            test_mountpoint.remove()
+            return result
+
+        # This is information about the dataset that is intentionally only
+        # populated once the path of the dataset is known by the cluster and
+        # thus could be known by an API client.
+        dataset_info_from_cluster = {
```

Comment:
Yea... the thing is I want to try to write to paths that are no longer being reported by flocker as valid paths for the dataset (Once flocker reports a path as a location that a dataset is located at, I want to keep trying to write there even if flocker is no longer reporting that as a valid path for the dataset) AFAIK this is similar to protecting against docker trying to use a path after flocker has removed it from a node for whatever reason (stale caches / configurations / etc).

I refactored this into an external agent class which hopefully more clearly shows that this is information that is known by some external agent (i.e. potential views of the world that docker may have, that are intentionally arbitrarily stale).